app/helpers/number_format_helper.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module NumberFormatHelper
+  def number_with_limit(number, opts = {})
+    init_formatting_options(opts) => {
+      delimiter:,
+      separator:,
+      digits:,
+      precision:,
+      length_limit:
+    }
+
+    string_number = number_with_precision(number, precision:, strip_insignificant_zeros: true)
+    length = string_number.delete("#{delimiter}#{separator}").length
+
+    scientific_notation_needed = length > length_limit ||
+                                 integer_part_size(number) > digits ||
+                                 (string_number == "0" && number != 0)
```

Comment:
This looks like a very specific edge case, I guess for very small numbers that get rendered as "0".

Do we have a proper test for this case?  I can't identify one.

What I'd expect to be tested:

* "0.0" (a floating point zero) can still be rendered as "0" very normally
* a number for which this rule was added and that's not caught by other rules is rendered scientifically

Reply:
actually this case is very common. It happens everytime, the precision is smaller, then the exponent (abs) for a small number.

`number_with_precision` will always render 0, if within the precision only zeros are found.